scripts/validate-autoinstall-user-data.py

```diff
@@ -27,64 +27,224 @@
-        assert user_data.readline() == "#cloud-config\n"
-        def get_autoinstall_data(data): return data["autoinstall"]
-    else:
-        def get_autoinstall_data(data):
-            try:
-                cfg = data["autoinstall"]
-            except KeyError:
-                cfg = data
-            return cfg
+    # Supress start and finish events unless verbosity >=2
+    if verbosity < 2:
+        for el in app.event_listeners:
+            el.report_start_event = lambda x, y: None
+            el.report_finish_event = lambda x, y, z: None
+    # Suppress info events unless verbosity >=1
+    if verbosity < 1:
+        for el in app.event_listeners:
+            el.report_info_event = lambda x, y: None
 
+    # Tell the server where to load the autoinstall
+    app.autoinstall = cfg_path
+    # Make sure events are printed (we could fail during read, which
+    # would happen before we setup the reporting controller)
+    app.controllers.Reporting.config = {"builtin": {"type": "print"}}
+    app.controllers.Reporting.start()
+    # Do both validation phases
+    try:
+        app.load_autoinstall_config(only_early=True, context=None)
+        app.load_autoinstall_config(only_early=False, context=None)
+    except Exception as exc:
 
-    # Verify autoinstall doc link is in the file
+        print(exc)  # Has the useful error message
+
+        # Print the full traceback if verbosity >=2
+        if verbosity > 2:
+            traceback.print_exception(exc)
+
+        print(FAILURE_MSG)
+        return 1
 
-    stream_pos: int = user_data.tell()
+    print(SUCCESS_MSG)
+    return 0
 
-    data: str = user_data.read()
 
-    link: str = "https://canonical-subiquity.readthedocs-hosted.com/en/latest/reference/autoinstall-reference.html"  # noqa: E501
+def main() -> int:
+    """Entry point."""
 
-    assert link in data
+    args: Namespace = parse_args()
+
+    user_data: io.TextIOWrapper = args.input
+    str_data: str = user_data.read()
+
+    # Verify autoinstall doc link is in the file
+    if args.check_link:
+        link: str = (
+            "https://canonical-subiquity.readthedocs-hosted.com/en/latest/reference/autoinstall-reference.html"  # noqa: E501
+        )
+
+        if link not in str_data:
+            raise AssertionError("Documentation link missing from user data")
+
+    # Parse out the autoinstall if expected within cloud-config
+    if args.expect_cloudconfig:
+        try:
+            ai_dict: dict[str, Any] = parse_cloud_config(str_data)
+        except Exception as exc:
+            print(f"{type(exc).__name__}: {exc}")
+            print(FAILURE_MSG)
+            sys.exit(1)
+
+        ai_data: str = yaml.dump(ai_dict)
+    else:
+        ai_data = str_data
 
-    # Verify autoinstall schema
-    user_data.seek(stream_pos)
+    with tempfile.TemporaryDirectory() as td:
+        path = Path(td) / "autoinstall.yaml"
+        with open(path, "w") as tf:
+            tf.write(ai_data)
```

Comment:
suggestion: 

```python
path.write_text(ai_data)
```

https://docs.python.org/3/library/pathlib.html#pathlib.Path.write_text

Reply:
TIL. Thanks for pointing that out!